library/serialize.c

```diff
@@ -0,0 +1,325 @@
+
+    if( ( ret = mbedtls_serialize_write( header, sizeof( header ) ) ) != 0 )
+        return( ret );
+    if( ( ret = mbedtls_serialize_write( buffer, length ) ) != 0 )
+        return( ret );
+    return( 0 );
+}
+
+int mbedtls_serialize_push_int16( uint16_t value )
+{
+    uint8_t buffer[2];
+    buffer[0] = ( value >> 8 ) & 0xff;
+    buffer[1] = value & 0xff;
+    return( mbedtls_serialize_push_buffer( buffer, sizeof( buffer ) ) );
+}
+
+int mbedtls_serialize_push_int32( uint32_t value )
+{
+    uint8_t buffer[4];
+    buffer[0] = ( value >> 24 ) & 0xff;
+    buffer[1] = ( value >> 16 ) & 0xff;
+    buffer[2] = ( value >> 8 ) & 0xff;
+    buffer[3] = value & 0xff;
+    return( mbedtls_serialize_push_buffer( buffer, sizeof( buffer ) ) );
+}
+
+int mbedtls_serialize_execute( uint32_t command )
+{
+    int ret;
+    uint8_t buffer[4];
+    buffer[0] = MBEDTLS_SERIALIZE_TYPE_EXECUTE;
+    buffer[1] = ( command >> 16 ) & 0xff;
+    buffer[2] = ( command >> 8 ) & 0xff;
+    buffer[3] = command & 0xff;
+    if( ( ret = mbedtls_serialize_write( buffer, sizeof( buffer ) ) ) != 0 )
+        return( ret );
+    if( ( ret = mbedtls_serialize_read( buffer, sizeof( buffer ) ) ) != 0 )
+        return( ret );
+    if( ! ( buffer[0] == MBEDTLS_SERIALIZE_TYPE_RESULT &&
+            buffer[1] == 0 && buffer[2] == 0 && buffer[3] == 4 ) )
+        return( MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT );
+    if( ( ret = mbedtls_serialize_read( buffer, sizeof( buffer ) ) ) != 0 )
+        return( ret );
+    return( buffer[0] << 24 | buffer[1] << 16 | buffer[2] << 8 | buffer[3] );
+}
+
+int mbedtls_serialize_pop_buffer( void *buffer, size_t max_length, size_t *actual_length )
+{
+    int ret;
+    uint8_t header[4];
+    if( ( ret = mbedtls_serialize_read( header, sizeof( header ) ) ) != 0 )
+        return( ret );
+    *actual_length = header[1] << 16 | header[2] << 8 | header[3];
+    if( *actual_length > max_length )
+        return( MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT );
+    if( ( ret = mbedtls_serialize_read( buffer, *actual_length ) ) != 0 )
+        return( ret );
+    return( 0 );
+}
+
+int mbedtls_serialize_pop_int16( uint16_t *value )
+{
+    int ret;
+    uint8_t buffer[2];
+    size_t length;
+    if( ( ret = mbedtls_serialize_pop_buffer( buffer, sizeof( buffer ),
+                                              &length ) ) != 0 )
+        return( ret );
+    if( length != sizeof( buffer ) )
+        return( MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT );
+    *value = buffer[0] << 8 | buffer[1];
+    return( 0 );
+}
+
+int mbedtls_serialize_pop_int32( uint32_t *value )
+{
+    int ret;
+    uint8_t buffer[4];
+    size_t length;
+    if( ( ret = mbedtls_serialize_pop_buffer( buffer, sizeof( buffer ),
+                                              &length ) ) != 0 )
+        return( ret );
+    if( length != sizeof( buffer ) )
+        return( MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT );
+    *value = ( buffer[0] << 24 ) | ( buffer[1] << 16 ) | ( buffer[2] << 8 ) | buffer[3];
```

Comment:
Same as above: please add an explicit cast to `uint32_t` before applying `<< 24`.